plugins/filter_aws/aws.c

```diff
@@ -427,8 +429,113 @@ static int get_ec2_tag_values(struct flb_filter_aws *ctx)
     return 0;
 }
 
+static int tag_is_present_in_list(struct flb_filter_aws *ctx, flb_sds_t tag,
+        const char *tags)
+{
+    const char *tag_start;
+    const char *tag_end;
+    const char *tag_sep_before;
+    const char *tags_start;
+    const char *tag_sep_after;
+    const char *tags_end;
+
+    tag_start = strstr(tags, tag);
+    if (!tag_start) {
+        return FLB_FALSE;
+    }
+    tag_end = tag_start + strlen(tag) - 1;
+
+    /* if substring exists it not necessarily implies that tag is included */
+    /* tag name must be exactly equal to the value between separators ',' */
+    /* f.e. without additional conditions, substring would accept Name for a,Namee,b */
+
+    tags_start = tags;
+    tags_end = tags + strlen(tags) - 1;
+
```

Comment:
I think this code is not intended because without it I think the logic makes sense to me.

Reply:
It is intended.

`tags_start` and `tags_end` define memory bounds for the `tags`. `tag_start` and `tag_end` define potential memory bounds of the `tag` value within the `tags` list.

Tag is present in the list only if:
 1. Tag is the first value in the tags list OR left tag memory bound is after a semicolon,
 2. Tag is the last value in the tags list OR right tag memory bound is before a semicolon,

In order to implement this in a general way, we always try to compute the positions of the semicolons, and if these values are outside of the `tags`, then we correct them to NULL. These are represented by `tag_sep_before` and `tag_sep_after`.

---

For example, if we would want to compute the function for the following arguments:

Tags: `Tag1,Tag2`
Tag: `Tag1`

strstr would find the substring in the tags list, the return value is the address of the first character of the substring.
tag_start = 0 // first character
tag_end = 3  // last character of the substring

tags_start = 0 // first character of tags list
tags_end = 8 // last character of tags list, but not \0

tag_sep_before = NULL // as tag_start - 1 is outside of the valid memory bounds
tag_sep_after = 4 // as it's tag_end + 1, it's not NULL, therefore we are going to expect there is a semicolon

Then we will check for the actual conditions,
```c
if (tag_start != tags_start && (tag_sep_before && *tag_sep_before != ',')) {
        return FLB_FALSE;
}
if (tag_end != tags_end && (tag_sep_after && *tag_sep_after != ',')) {
        return FLB_FALSE;
}
```

First `if` won't return false, as `tag_sep_before` is NULL. Second `if` won't return false, as `tag_sep_after` points to the semicolon.

In the end, the function will return true.

 - Computation for the `Tag2` is equivalent,
 - If `tags` won't have the `tag` substring, it will return false right away,
 - If `tags` will have the `tag` substring, then it won't be surrounded by semicolons, and we will also return false;

Reply:
I am open to any suggestions how we could simplify this function. I tried my best to implement the simplest version, but of course there might be some area for improvement.

Also, feel free to suggest any improvements in the comments, such that it makes the implementation easier to understand.

Reply:
(Actually the comparisons `tag_start != tags_start` and `tag_end != tags_end` are not needed.)

Reply:
As noted in another comment, why not turn `tags_include` into a nice list first, same as tag_keys is a nice list ?

Reply:
This is an old implementation, which no longer exists, so I am going to resolve the conversation.